File: src/09_Range_Queries/sum_queries_1d.cpp
#include <bits/stdc++.h>
#include "../compete.hpp"
using namespace std;


vector<int> v{1, 3, 4, 8, 6, 1, 4, 2};
int q;
int a, b;


void solve()
{
  // cumulative vector
  vector<int> cumsum(v.size());
  partial_sum(v.begin(), v.end(), cumsum.begin());

  cout << "Give number of queries: ";
  cin >> q;

  cout << v << '\n';

  while (q--) {
    cout << "Give subset limits: ";
    cin >> a >> b;
    cout << "Subset sum: " << cumsum[b] - cumsum[a] + v[a] << '\n';
  }
}


int main() {solve();}

File: src/09_Range_Queries/sum_queries_1d.py
from itertools import accumulate


v = [1, 3, 4, 8, 6, 1, 4, 2]
cumsum = list(accumulate(v))


def solve(a, b):
  return cumsum[b] - cumsum[a] + v[a]


if __name__ == '__main__':
  queries = (
    ((1, 3), 15),
    ((0, len(v) - 1), sum(v)),
    ((0, 3), 16),
    ((4, 7), 13),
    ((5, 6), 5)
  )

  for q, expected in queries:
    ans = solve(*q)
    print(ans)
    assert ans == expected
